File: src/main/java/com/qtech/bubbles/registry/Registers.java
package com.qtech.bubbles.registry;

import com.google.common.annotations.Beta;
import com.qtech.bubbles.bubble.AbstractBubble;
import com.qtech.bubbles.common.ResourceEntry;
import com.qtech.bubbles.common.effect.StatusEffect;
import com.qtech.bubbles.common.gamestate.GameEvent;
import com.qtech.bubbles.common.gametype.AbstractGameType;
import com.qtech.bubbles.entity.ammo.AmmoType;
import com.qtech.bubbles.entity.player.ability.AbilityType;
import com.qtech.bubbles.entity.types.EntityType;
import com.qtech.bubbles.graphics.TextureCollection;
import com.qtech.bubbles.gui.cursor.RegistrableCursor;
import com.qtech.bubbles.init.*;
import com.qtech.bubbles.item.ItemType;

/**
 * Registers, use this class to create deferred registers for specific things in the game.
 *
 * @see DeferredRegister
 */
@SuppressWarnings({"rawtypes", "deprecation"})
public class Registers {
    /**
     * Bubble register.
     * Bubbles are the core mechanic of the game, without the bubbles the game shouldn't exist.
     *
     * @see Bubbles
     */
    public static final Registry<AbstractBubble> BUBBLES = Registry.create(AbstractBubble.class, new ResourceEntry("qbubbles", "bubbles"));

    /**
     * Ammo type register.
     * Ammo types are what it says ammo types, you can create different ammo types doing different things.
     *
     * @see AmmoTypes
     */
    public static final Registry<AmmoType> AMMO_TYPES = Registry.create(AmmoType.class, new ResourceEntry("qbubbles", "ammo_types"));

    /**
     * Entity register.
     * Entities are one of the core things in a game, they are used for creating the player, enemies or in this case also bubbles.
     *
     * @see Entities
     */
    public static final Registry<EntityType> ENTITIES = Registry.create(EntityType.class, new ResourceEntry("qbubbles", "entities"));

    /**
     * Effect register.
     * Effects are classes for doing things over time after activating, and can stop after some time.
     *
     * @see Effects
     */
    public static final Registry<StatusEffect> EFFECTS = Registry.create(StatusEffect.class, new ResourceEntry("qbubbles", "effects"));

    /**
     * Ability register.
     * Abilities are classes for doing things like teleportation.
     *
     * @see Abilities
     */
    public static final Registry<AbilityType> ABILITIES = Registry.create(AbilityType.class, new ResourceEntry("qbubbles", "abilities"));

    /**
     * Game state register.
     *
     * @see GameEvents
     */
    public static final Registry<GameEvent> GAME_EVENTS = Registry.create(GameEvent.class, new ResourceEntry("qbubbles", "game_states"));

    /**
     * Game type register.
     *
     * @see GameTypes
     */
    public static final Registry<AbstractGameType> GAME_TYPES = Registry.create(AbstractGameType.class, new ResourceEntry("qbubbles", "game_types"));

    /**
     * Cursor register.
     */
    public static final Registry<RegistrableCursor> CURSORS = Registry.create(RegistrableCursor.class, new ResourceEntry("qbubbles", "cursors"));

    /**
     * Items register, will be used in the future.
     */
    @Beta
    public static final Registry<ItemType> ITEMS = Registry.create(ItemType.class, new ResourceEntry("qbubbles", "items"));

    /**
     * Texture collection register.
     * Used for mapping and preloading textures.
     *
     * @see TextureCollections
     */
    public static final Registry<TextureCollection> TEXTURE_COLLECTIONS = Registry.create(TextureCollection.class, new ResourceEntry("qbubbles", "texture_collections"));
}

File: src/main/kotlin/com/qtech/bubbles/registry/Registers.kt
package com.qtech.bubbles.registry

import com.google.common.annotations.Beta
import com.qtech.bubbles.bubble.AbstractBubble
import com.qtech.bubbles.common.ResourceLocation
import com.qtech.bubbles.common.ability.AbilityType
import com.qtech.bubbles.common.ammo.AmmoType
import com.qtech.bubbles.common.cursor.RegistrableCursor
import com.qtech.bubbles.common.effect.Effect
import com.qtech.bubbles.common.gamestate.GameEvent
import com.qtech.bubbles.common.gametype.AbstractGameMode
import com.qtech.bubbles.entity.types.EntityType
import com.qtech.bubbles.graphics.TextureCollection
import com.qtech.bubbles.item.Item

/**
 * Registers, use this class to create deferred registers for specific things in the game.
 *
 * @see DeferredRegister
 */
object Registers {
    /**
     * Bubble register.
     * Bubbles are the core mechanic of the game, without the bubbles the game shouldn't exist.
     *
     * @see Bubbles
     */
    val BUBBLES: Registry<AbstractBubble> = Registry.create(AbstractBubble::class.java, ResourceLocation("bubbleblaster", "bubbles"))

    /**
     * Ammo type register.
     * Ammo types are what it says ammo types, you can create different ammo types doing different things.
     *
     * @see AmmoTypes
     */
    val AMMO_TYPES: Registry<AmmoType> = Registry.create(AmmoType::class.java, ResourceLocation("bubbleblaster", "ammo_types"))

    /**
     * Entity register.
     * Entities are one of the core things in a game, they are used for creating the player, enemies or in this case also bubbles.
     *
     * @see Entities
     */
    val ENTITIES: Registry<EntityType<*>> = Registry.create(EntityType::class.java, ResourceLocation("bubbleblaster", "entities"))

    /**
     * Effect register.
     * Effects are classes for doing things over time after activating, and can stop after some time.
     *
     * @see Effects
     */
    val EFFECTS: Registry<Effect> = Registry.create(Effect::class.java, ResourceLocation("bubbleblaster", "effects"))

    /**
     * Ability register.
     * Abilities are classes for doing things like teleportation.
     *
     * @see Abilities
     */
    val ABILITIES: Registry<AbilityType<*>> = Registry.create(AbilityType::class.java, ResourceLocation("bubbleblaster", "abilities"))

    /**
     * Game state register.
     *
     * @see GameEvents
     */
    val GAME_EVENTS: Registry<GameEvent> = Registry.create(GameEvent::class.java, ResourceLocation("bubbleblaster", "game_states"))

    /**
     * Game type register.
     *
     * @see GameTypes
     */
    val GAME_TYPES: Registry<AbstractGameMode> = Registry.create(AbstractGameMode::class.java, ResourceLocation("bubbleblaster", "game_types"))

    /**
     * Cursor register.
     */
    val CURSORS: Registry<RegistrableCursor> = Registry.create(RegistrableCursor::class.java, ResourceLocation("bubbleblaster", "cursors"))

    /**
     * Items register, will be used in the future.
     */
    @Beta
    val ITEMS: Registry<Item> = Registry.create(Item::class.java, ResourceLocation("bubbleblaster", "items"))

    /**
     * Texture collection register.
     * Used for mapping and preloading textures.
     *
     * @see TextureCollections
     */
    val TEXTURE_COLLECTIONS: Registry<TextureCollection> = Registry.create(TextureCollection::class.java, ResourceLocation("bubbleblaster", "texture_collections"))
}